### ISSUE TITLE
PartitionedOutput operator improvement 1: reduce blocking

### ISSUE THREAD

**Issue by yingsu00** [facebookincubator/velox#10608]
### Description

Our internal performance testing shows that the PartitionedOutput performance was not as desired. The problem of it lies in 3 areas:
1. High blocking time
2. High CPU time
3. High memory usage

To illustrate 1, we found the ExchangeOperator is constantly blocked and waiting for the PartitionedOutput operator to fill the output buffers. This made ExchangeOperator the top operator in blocking time in TPCDS runs.
<img width="491" alt="Screenshot 2024-07-30 at 11 00 44" src="https://github.com/user-attachments/assets/7ae24a74-12b5-47d2-a1a3-371379df95d4">
I also believe it is one of the main reasons that Velox wall time improvement is not as good as CPU time improvement. In TPCH tests, the CPU of Presto C++ improved 3x comparing to Presto Java, but the wall time only improved 2x. And on a cluster that is not blocked on IO and with using full number of cores, Presto C++ could never saturate the CPU. 

To illustrate 2, in TPCH 100GB, it is the 2nd highest operator in all operators in terms of CPU time, and is over 2x regression comparing to Presto Java OptimizedPartitionedOutputOperator.
<img width="421" alt="image" src="https://github.com/user-attachments/assets/64df22ed-196b-4a46-aa4e-143b59be09f7">

To illustrate 3,  in TPCH 100GB, the memory usage of this operator is 4x more than Presto Java OptimizedPartitionedOutputOperator.
<img width="825" alt="image" src="https://github.com/user-attachments/assets/9177d8f7-1a64-40cf-941f-313121778379">

In this Github issue, we discuss how to solve 1. I also have some ideas how to improve 2 and 3 so as to achieve even better results than Presto Java OptimizedPartitionedOutputOperator, but that will be the second step after we solve the blocking issue.

The blocking issue is caused by several things:
1. Retrieving the OutputBuffer for each task is under a global lock. In Velox, the output buffers were held in a std::unordered_map of <TaskId, OutputBuffer> in the singleton OutputBufferManager, and is guarded by a std::mutex. Even though the task is already identified when receiving a request from the coordinator, it still need to lookup this <TaskId, OutputBuffer> map to get the OutputBuffer pointer. This look up is guarded by the mutex and is mutually exclusive, even though the Task - OutputBuffer mapping is fixed.
```
  folly::Synchronized<
      std::unordered_map<std::string, std::shared_ptr<OutputBuffer>>,
      std::mutex> // std::mutex instead of std::shared_mutex, such that withRLock cannot be used
      buffers_;
```
2. All operations like enqueue, getData, ack, etc. are guarded by a mutex at the OutputBuffer level. But in fact, for an OutputBuffer of partitioned type, the vector of DestinationBuffers is immutable and taking an exclusive lock on the whole vector for enqueueing/dequeueing a SerializedPage to only one DestinationBuffers is not necessary.

In my upcoming PRs I will do the following:
1. Reduce the critical section for enqueueing and dequeueing to the DestinationBuffer level whenever possible.
2. Replace the <TaskId, OutputBuffer> with ConcurrentHashMap, in which readers are always wait-free, and writes are high performance. Replace the counters to atomic such that updating them does not need to be under big exclusive locks.
3. In the next step, make the Task directly own the OutputBuffer, such that the lookup of OutputBuffer is removed.

**Comment by yingsu00** [comment]
cc @pedroerp @Yuhta I have the PR almost ready and is doing performance testing.

**Comment by Yuhta** [comment]
Shuffle change must be tested in shadow with real workloads (~1k queries at least).  Do you have the environment for such testing?

**Comment by pedroerp** [comment]
Cc: @mbasmanova @xiaoxmeng @oerling

**Comment by yingsu00** [comment]
> Shuffle change must be tested in shadow with real workloads (~300 nodes ~1k queries with 10 queries running concurrently at least). Have you prepared such testing (we can get some wall time improvement metrics with that setup as well)?

@Yuhta  We can probably shadow some customer workloads who uses big clusters, in addition to TPCDS 100TB throughput test (multiple streams) on ~100 nodes. As the first step I'll start with smaller workloads and clusters.

But nonetheless, I understand there might be risk just enabling the new code by default. Considering I also want to improve the CPU time and memory usage of this operator and will touch more code, how about I make the new code under some config flag that can be turned on/off easily? Then you can try it in your internal environment and roll out step by step. I can closely work with you to debug and resolve problems that you might find.

btw I don't quite understand "(we can get some wall time improvement metrics with that setup as well)?", could you explain a bit more? Thanks!

**Comment by Yuhta** [comment]
Yeah a config to guard the new code would be helpful, this way we can disable it in prod while enable it in some shadow. Running heterogeneous workloads instead of TPC would also be helpful.  If you have a way to replay same workload on 2 different clusters, one running old code and one running new code, you can compare the wall time of the 2 and get a feeling of how the new code improve wall time.  This is actually the only way to measure wall time improvements for such changes.

**Comment by mbasmanova** [comment]
CC: @arhimondr

**Comment by kevinwilfong** [comment]
I've looked at improving the CPU time of PartitionedOutput operator in the past and it looks like replicating Presto's TaskOutput operator for cases where we're we don't need to really shuffle the data (e.g. round robin, broadcast) could help.

https://github.com/facebookincubator/velox/pull/8740

I was looking at it again this week and PrestoBatchVectorSerializer's performance is still not where I want it, it does particularly bad on very wide rows, and that's something I'm working on this week before I land the above PR.

**Comment by yingsu00** [comment]
> @yingsu00 thanks for starting this thread! I wonder if you could come up with a small benchmark that we could use to reproduce and iterate on these performance enhancements? What @majetideepak and others did for tpch was to create a few C++ binaries with the Velox plans hard-coded, so that we could easily reproduce issues and iterate on performance.
> 
> In case we haven't them already, could you add a few tpcds examples that we could use to reproduce this behavior?

Hi @pedroerp yes I will include some microbenchmark. I was thinking of generating multiple threads and random data. We will also add Presto C++ e2e benchmarks. Your suggested benchmark seems to be the counter part of TPCH. We can add it in the next step.

I'm currently testing my code with small clusters and found some issue. I'm trying to fix it via several different ways and will pick the best one. Will post update later.
